### ISSUE TITLE
 exception: std::logic_error

### ISSUE THREAD

**Issue by FabioMBB** [mlpack/examples#144]
Hi, 
I'm trying to run the mnist_cnn example but I keep getting the following error:

Reading data ...
Start training ...

error: Mat::operator(): index out of bounds
terminate called after throwing an instance of 'std::logic_error'
  what():  Mat::operator(): index out of bounds

Is it ever happen to any of you?

Thank you so much in advance for your help,
Fabio

**Comment by FabioMBB** [comment]
NVM I had to convert the dataset with the supplied python function instead of using the original one. 
Thank you anyway.

**Comment by FabioMBB** [comment]
In reality I had the issue is still not solved...
To solve it I had to add 1 to the y values and normalize the X.

**Comment by rcurtin** [comment]
@FabioMBB glad you got it figured out---sorry we weren't able to help you sooner.  What version of mlpack are you using?  We just merged changes to master (and this branch) in mlpack/mlpack#2534 and #142, so if you aren't using mlpack git master then you will indeed need to add 1 to `y`.  If normalizing the data significantly helps convergence, want to open a PR?

**Comment by FabioMBB** [comment]
Thank you very much rcurtin. I'm using mlpack-master. But maybe I have to update it.

Anyway, the results are still strange. Even when I normalize the input, the error in validation is very high. I'm not sure what I'm doing wrong....

**Comment by FabioMBB** [comment]
The modifications I did are the following ones: 
 // Create labels for training and validatiion datasets.
  mat trainY = train.row(0)+1;
  mat validY = valid.row(0)+1;
  
  // Scale all data into the range (0, 1) for increased numerical stability.
  data::MinMaxScaler scaleX;
  // Fit scaler only on training data.
  scaleX.Fit(trainX);
  scaleX.Transform(trainX, trainX);
  scaleX.Transform(validX, validX);

**Comment by FabioMBB** [comment]
The modifications I did are the following ones:
// Create labels for training and validatiion datasets.
  mat trainY = train.row(0)+1;
  mat validY = valid.row(0)+1;
  
  // Scale all data into the range (0, 1) for increased numerical stability.
  data::MinMaxScaler scaleX;
  // Fit scaler only on training data.
  scaleX.Fit(trainX);
  scaleX.Transform(trainX, trainX);
  scaleX.Transform(validX, validX);

**Comment by rcurtin** [comment]
@FabioMBB can you tell me what the validation error you are seeing is?  When I run against mlpack master (so that no changes are needed to `mnist_cnn.cpp` for labels), I get this:

```
Validation loss: 20273.7.
Accuracy: train = 99.3519%,	 valid = 97.4996%
```

This seems like about what I would expect.  Are you seeing a significantly different result?

**Comment by FabioMBB** [comment]
@rcurtin Hi Ryan, thank you so much for following up on this. I just figured out my problem. I did not add a +1 also in the prediction labels. That was causing me to have the wrong labels with respect to the validation set. I should have updated the code with the last repository before running this example to avoid these problems. Anyway I really appreciated your help.

**Comment by rcurtin** [comment]
Awesome!  Well I also found a minor bug in the implementation (it does not really affect accuracy :)) so I'll have a PR open for that once I finish local testing.  Thanks for opening the issue!  More eyes on code generally mean more issues get caught. :)

**Comment by FabioMBB** [comment]
Hi Ryan,
Sorry to bother you again,
But I’m having some trouble with the CNN running in parallel cores.

When I use a batch of 60, all processors run in parallel at 100%, but if I increase the batch size to 6000, I get one core at 100% and the others at 20%.
I need large batches for the project I’m using MLPACK.

Do you know why this is happening?

Thank you so much in advance for your help,
Best Regards,
Fabio

**Comment by rcurtin** [comment]
We primarily expect OpenBLAS to provide the parallelism, so at least from our side I'm not sure there's too much we can do.  One thing to check would be that you are using OpenBLAS, and that you are compiling with `-fopenmp` (but I would suspect that because you are seeing any parallel behavior at all, you are compiling with OpenMP support already).  OpenBLAS may not be able to parallelize as well depending on the nature of the linear algebra operations that are running.

You might consider playing with the `OMP_NUM_THREADS` environment variable when running the code to see if you can get it to max out a fewer number of cores.  Sometimes due to communication overhead between threads and perhaps contention issues, it can be faster to run with fewer cores overall.

**Comment by FabioMBB** [comment]
Thank you very much Ryan for your response.
Yes, I’m using -fopenmp and using OpenBLAS.

I was actually able to get all the cores running near 80% by using the work done by  adiwajshing (https://github.com/mlpack/mlpack/pull/2341).

I also tried to speed up the for loop in NaiveConvolution as follow (adding the intermediate variables, so that they get calculated only once), but that does not seem to make any difference, not sure if the compiler already takes care of the intermediate calculation in a very efficient way:


#pragma omp parallel for schedule(static) if (input.n_cols >= 64)
    for (omp_size_t j = 0; j < o_cols; ++j)
    {
j_dW=j*dW;
j_o_rows=j*o_rows;
      for (size_t i = 0; i < o_rows; ++i)
      {
i_dH=i*dH;
j_o_rows_i=j_o_rows+i;
               const eT* kernelPtr = filter.memptr();
kj_dilationW=j_dW;
        for (size_t kj = 0; kj < f_cols; ++kj)
        {
               const eT* inputPtr = input.colptr(kj_dilationW) + i_dH;
kj_dilationW+=dilationW;

         for (size_t ki = 0; ki < f_rows; ++ki, ++kernelPtr,
              inputPtr += dilationH)
           outputPtr[j_o_rows_i] += *kernelPtr * (*inputPtr);
        }
      }
    }

Thank you so much for all you are doing,
Best regards,
Fabio

**Comment by rcurtin** [comment]
Nice, thank you for the information!  Let's see if we can bring #2341 back to life. :+1:

Did you find that with the modifications, that the overall computation ran faster too?

**Comment by FabioMBB** [comment]
Hi Ryan,
The modifications do not seem to speed up the computation, which is very strange. I will run some more test and update you soon.

Thank you,
Fabio

**Comment by FabioMBB** [comment]
Hi Ryan,
The last version of the loops I’m testing in naïve_convolutional are the following two:

* Perform a convolution (valid mode).

#pragma omp parallel for schedule(static) if (input.n_cols >= 64)
    for (omp_size_t j = 0; j < o_cols; ++j)
    {
     size_t j_dW=j*dW;         //this is done as a product to avoid interference with OMP
     size_t j_o_rows=j*o_rows; //this is done as a product to avoid interference with OMP

      for (size_t i = 0, j_o_rows_i = j_o_rows, i_dH = 0; i < o_rows; ++i, j_o_rows_i++, i_dH += dH)
      {
        const eT* kernelPtr = filter.memptr();
        for (size_t kj = 0, kj_dilationW = j_dW; kj < f_cols; ++kj, kj_dilationW += dilationW)
        {
         const eT* inputPtr = input.colptr(kj_dilationW) + i_dH;

         for (size_t ki = 0; ki < f_rows; ++ki, ++kernelPtr, inputPtr += dilationH)
           outputPtr[j_o_rows_i] += *kernelPtr * (*inputPtr);
        }
      }
    }

* Perform a convolution (full mode).

size_t filter_n_rows_1_dilationW=(filter.n_rows - 1) * dilationW;
    size_t filter_n_cols_1_dilationH=(filter.n_cols - 1) * dilationH;
    size_t outputRows = (input.n_rows - 1) * dW + 2 * filter_n_rows_1_dilationW + 1;
    size_t outputCols = (input.n_cols - 1) * dH + 2 * filter_n_cols_1_dilationH + 1;
    size_t outputRows_filter_n_rows_dilationW=outputRows - 2 * filter_n_rows_1_dilationW - 1;
    size_t outputCols_filter_n_cols_dilationH=outputCols - 2 * filter_n_cols_1_dilationH - 1;
    for (size_t i = 0; i < dW; i++)
    {
      if (((((i + outputRows_filter_n_rows_dilationW) % dW) + dW) % dW) == i){
        outputRows += i;
        break;
      }
    }
    for (size_t i = 0; i < dH; i++)
    {
      if (((((i + outputCols_filter_n_cols_dilationH) % dH)
          + dH) % dH) == i){
        outputCols += i;
        break;
      }
    }

However, I do not see much difference in performance, maybe you or @adiwajshing<https://github.com/adiwajshing> know why?

Please advise,
Thank you very much,
Fabio

**Comment by rcurtin** [comment]
@FabioMBB it might be worth trying the whole change from mlpack/mlpack#2341.  I tried it locally, modifying the `mnist_cnn` example to only run for two epochs, and here's what I saw on `master`:

```
Accuracy: train = 88.3611%,...]  valid = 88%6s - loss: 186.523
Predicting ...
Saving predicted labels to "results.csv."...
Neural network model is saved to "model.bin"
Finished

real    2m7.439s
user    10m31.966s
sys     0m2.357s
```

And on #2341:

...
Accuracy: train = 86.237%,....]  valid = 86.1167%loss: 1744.42
Predicting ...
Saving predicted labels to "results.csv."...
Neural network model is saved to "model.bin"
Finished

real    0m42.981s
user    14m3.446s
sys     0m27.264s
```

So it seems like there is a significant improvement!  However, I think there is a minor issue in the computation, since I am seeing different results.  I'll reopen #2341 and see if we can start pushing towards incorporating it. :+1:

**Comment by FabioMBB** [comment]
Thank you very much Ryan for looking into this and glad to hear that we are getting a speed up.

Not sure about the difference in computation.

Did you try mlpack/mlpack#2341<https://github.com/mlpack/mlpack/pull/2341> with or without the changes on the loops?

Please let me know,
Thank you very much,
Fabio

**Comment by rcurtin** [comment]
I used all the changes from the branch, so maybe it is worth trying that on your end to see what happens?

**Comment by FabioMBB** [comment]
Thank you very much Ryan for letting me know.

Yes, as soon as @adiwajshing finishes his review I will run some tests.

Thank you again,
Fabio
